libamqpprox/amqpprox_connectionlimitermanager.cpp

```diff
@@ -0,0 +1,266 @@
+    return connectionRateLimiter;
+}
+
+std::shared_ptr<ConnectionLimiterInterface>
+ConnectionLimiterManager::addAlarmOnlyConnectionRateLimiter(
+    const std::string &vhostName,
+    uint32_t           numberOfConnections)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    std::shared_ptr<FixedWindowConnectionRateLimiter>
+        alarmOnlyConnectionRateLimiter =
+            std::make_shared<FixedWindowConnectionRateLimiter>(
+                numberOfConnections);
+    d_alarmOnlyConnectionRateLimitersPerVhost[vhostName] = {
+        true, alarmOnlyConnectionRateLimiter};
+    return alarmOnlyConnectionRateLimiter;
+}
+
+void ConnectionLimiterManager::setDefaultConnectionRateLimit(
+    uint32_t defaultConnectionRateLimit)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultConnectionRateLimit = defaultConnectionRateLimit;
+    for (auto &limiter : d_connectionRateLimitersPerVhost) {
+        if (!limiter.second.first) {
+            limiter.second.second =
+                std::make_shared<FixedWindowConnectionRateLimiter>(
+                    d_defaultConnectionRateLimit);
+        }
+    }
+}
+
+void ConnectionLimiterManager::setAlarmOnlyDefaultConnectionRateLimit(
+    uint32_t defaultConnectionRateLimit)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultAlarmOnlyConnectionRateLimit = defaultConnectionRateLimit;
+    // To update new default connection rate limit for all the vhosts,
+    // by removing old already set default connection rate limiters
+    for (auto &limiter : d_alarmOnlyConnectionRateLimitersPerVhost) {
+        if (!limiter.second.first) {
+            limiter.second.second =
+                std::make_shared<FixedWindowConnectionRateLimiter>(
+                    d_defaultAlarmOnlyConnectionRateLimit);
+        }
+    }
+}
+
+void ConnectionLimiterManager::removeConnectionRateLimiter(
+    const std::string &vhostName)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_connectionRateLimitersPerVhost.erase(vhostName);
+    if (d_defaultConnectionRateLimit) {
+        d_connectionRateLimitersPerVhost[vhostName] = {
+            false,
+            std::make_shared<FixedWindowConnectionRateLimiter>(
+                d_defaultConnectionRateLimit)};
+    }
+}
+
+void ConnectionLimiterManager::removeAlarmOnlyConnectionRateLimiter(
+    const std::string &vhostName)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_alarmOnlyConnectionRateLimitersPerVhost.erase(vhostName);
+    if (d_defaultAlarmOnlyConnectionRateLimit) {
+        d_alarmOnlyConnectionRateLimitersPerVhost[vhostName] = {
+            false,
+            std::make_shared<FixedWindowConnectionRateLimiter>(
+                d_defaultAlarmOnlyConnectionRateLimit)};
+    }
+}
+
+void ConnectionLimiterManager::removeDefaultConnectionRateLimit()
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultConnectionRateLimit = 0;
+    for (auto it = d_connectionRateLimitersPerVhost.cbegin();
```

Comment:
take a look at [remove_if](https://en.cppreference.com/w/cpp/algorithm/remove), you can use this with `equal_range` this would be a 1 liner.

Reply:
As mentioned earlier, here we are iterating based on the specific value condition, instead of specific key.

Reply:
you're right, equal_range cant be used here cause its conditional on the value, but the remove_if algo is still valid if you have a suitable predicate function, but this is acceptable

Reply:
I think, `remove_if` isn't available for `unordered_map`. There is a similar function `erase_if` for `unordered_map`. But it is available from C++20. And in that, we don't have to worry about maintaining valid iterator with visiting and erasing elements all together.